### PR TITLE
[apiview] adding pr support

### PR DESCRIPTION
**Pull request by ckairen**
closes https://github.com/Azure/azure-sdk-tools/issues/8821

### CODE REVIEW

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (praveenkuttappan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:356]: `Detect-API-changes` step should only run after publishing the artifacts. This step sends a request to APIView and APIView will try to access the published artifact when processing the request. If artifact is not published, then request will fail with 404 error.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (praveenkuttappan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:359]: This step should only be triggered if Build.Reason is PullRequest.

File: eng/scripts/Prepare-APIReview.ps1
Comment (praveenkuttappan) [inline eng/scripts/Prepare-APIReview.ps1:11]: You don't need these args to generate token file. It's only required to trigger `Create-APIReview.ps1`

File: eng/scripts/Prepare-APIReview.ps1
Comment (praveenkuttappan) [inline eng/scripts/Prepare-APIReview.ps1:19]: Same here. These are not required to generate token files.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (praveenkuttappan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:134]: Step should be named as something like `Generate API review token files`

Comment (ckairen) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:134]: and trim down on the parameters

File: eng/scripts/Generate-APIReview-Token-Files.ps1
Comment (benbp) [inline eng/scripts/Generate-APIReview-Token-Files.ps1:28]:
```suggestion
    Write-Host "Creating API review artifact for $ArtifactName"
    New-Item -ItemType Directory -Path "$OutPath/$ArtifactName" -Force
```

File: eng/scripts/Generate-APIReview-Token-Files.ps1
Comment (benbp) [inline eng/scripts/Generate-APIReview-Token-Files.ps1:32]:
```suggestion
    & $ParserPath -o "$OutPath/$ArtifactName/${ArtifactName}_cpp.json" $SourcePath
```

ArtifactName here needs to be surrounded with `${}` otherwise I think it will look for a variable `$ArtifactName_cpp`.

File: eng/scripts/Generate-APIReview-Token-Files.ps1
Comment (benbp) [inline eng/scripts/Generate-APIReview-Token-Files.ps1:30]: Nit:

```suggestion
    Write-Host "Contents in $parentPath:"
```

**Comment by benbp** [comment]
powershell/yaml LGTM but I'll defer to Praveen for approval.

File: sdk/template/azure-template/src/template_client.cpp
Comment (LarryOsterman) [inline sdk/template/azure-template/src/template_client.cpp:3]: I'm surprised this can trigger a code review update - the code review json files *should* be idempotent and the only lines that would change (the #line directives) should be marked as "exclude from diff".

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:103]: For the download you should be able to just use the pwsh task without a service connection.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:103]: We will want to build in the version in the path so we can get the exact version.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (maririos) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:332]:
```suggestion
          displayName: Copy Package Json
```

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (praveenkuttappan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:163]: Can you use a variable to set this version and use variable instead of version number directly here?

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (praveenkuttappan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:185]: Any specific reason to use force create directory?

Comment (ckairen) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:185]: so it won't error out if the directory is already created, ran into that error when testing around with the pipeline

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (praveenkuttappan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:209]: What does this step do? Why do we need to run this step to get package version? Isn't this something that's already done by `Dump Package properties` or can we include the output from line 201-202 in the package info file created in previous step? What we need to do is to include the steps in Language-Setings.ps1 like other languages do.

Comment (ckairen) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:209]: included in the language0settings.ps1

File: sdk/template/azure-template/src/template_client.cpp
Comment (ckairen) [inline sdk/template/azure-template/src/template_client.cpp:3]:
```suggestion
```

Comment (ckairen) [reply sdk/template/azure-template/src/template_client.cpp:3]: ***delete this before merging***

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:164]: We should not be pulling the lastest version automatically. The version we depend on should be hard coded and updated when we update the parser. We need to keep the version pinned to what is currently working in the repo.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (praveenkuttappan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:164]: It's a better practice to put this under variable in line 88 as job variable and refer it.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (praveenkuttappan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:349]:
```suggestion
                ArtifactPath: $(Build.ArtifactStagingDirectory)/packages
```

**Review by praveenkuttappan — APPROVED** [review]
Two minor comments, Otherwise it looks fine. Please remove test line added in template package.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (ckairen) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:349]: this one doesn't work because in the detect-api-changes, there's a place where it uses -replace artifactPath, and since we're using windows os, we'll need the backwards slash

**Comment by ckairen** [comment]
pipeline run: https://dev.azure.com/azure-sdk/public/_build/results?buildId=4186257&view=logs&j=a7bbfb1e-9b1d-5480-b954-b42b4949e5ba&t=d2882e8e-3f84-5d13-b2a9-0f13ca3ef387

pr comment generated by this step: https://github.com/Azure/azure-sdk-for-cpp/pull/5932#issuecomment-2363931671

**Comment by ckairen** [comment]
all of the azure-template changes are for testing purposes, will delete before merging

**Review by danieljurek — APPROVED** [review]
LGTM. Revert the changes to template package before merging.

File: sdk/template/azure-template/src/template_client.cpp
Comment (ckairen) [inline sdk/template/azure-template/src/template_client.cpp:71]:
```suggestion
```

File: sdk/template/azure-template/src/private/package_version.hpp
Comment (ckairen) [inline sdk/template/azure-template/src/private/package_version.hpp:44]:
```suggestion
```

File: sdk/template/azure-template/inc/azure/template/template_client.hpp
Comment (ckairen) [inline sdk/template/azure-template/inc/azure/template/template_client.hpp:40]:
```suggestion
```

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (benbp) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:165]: I don't think it will render properly with the `${{` compile time syntax.

```suggestion
            Write-Host "Using apiview parser version $(apiviewParserVersion)"
```

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml
Comment (benbp) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:166]: Same change as above.

**Pull request merged**
